Trilinos nightly failure, create_mirror_view incompatibilities with Sacado

**Describe the bug**

Nightly builds of Trilinos fail to compile with apparent incompatibilities following `create_mirror_view` updates with #7061

**Output snip**:
```
17:43:17 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/packages/sacado/test/UnitTests/Fad_Kokkos_CommTests_Serial.cpp:54:1: error: no matching function for call to 'create_mirror_view'
17:43:17 FAD_KOKKOS_COMM_TESTS_SERIAL(Fad_DFadType, Fad_DFad)
17:43:17 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17:43:17 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/packages/sacado/test/UnitTests/Fad_CommTests.hpp:1211:3: note: expanded from macro 'FAD_KOKKOS_COMM_TESTS_SERIAL'
17:43:17   FAD_KOKKOS_COMM_TESTS_DEV(FadType, FAD, Serial)
17:43:17   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17:43:17 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/packages/sacado/test/UnitTests/Fad_CommTests.hpp:930:22: note: expanded from macro 'FAD_KOKKOS_COMM_TESTS_DEV'
17:43:17   HostViewType h_x = Kokkos::create_mirror_view(x);                     \
17:43:17                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
17:43:17 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_CopyViews.hpp:3608:6: note: candidate template ignored: requirement 'std::is_void_v<Kokkos::Impl::ViewSpecializeSacadoFad>' was not satisfied [with T = Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> *, P = <Kokkos::Serial>]
17:43:17 auto create_mirror_view(const Kokkos::View<T, P...>& src) {
...
```

**Please include the following for a minimal reproducer**

**Reproducer configuration notes**:
The error occurs with all backends, below is a configuration with llvm/15 and Serial backend
```bash
# Repo prep
git clone -b develop https://github.com/trilinos/Trilinos.git
TRILINOS_DIR=$PWD/Trilinos
git clone -b develop https://github.com/kokkos/kokkos.git
KOKKOS_DIR=$PWD/kokkos
git clone -b develop https://github.com/kokkos/kokkos-kernels.git
KOKKOSKERNELS_DIR=$PWD/kokkos-kernels

cd $TRILINOS_DIR
ln -s ${KOKKOS_DIR}/kokkos kokkos
ln -s ${KOKKOSKERNELS_DIR}/kokkos-kernels kokkos-kernels
cd -

mkdir -p Build
cd Build

# Setup environment - Blake testbed "all" queue shown here
module load cmake llvm openblas/0.3.23

# Configure Trilinos
cmake \
  -D CMAKE_INSTALL_PREFIX="${TRILINOS_INSTALL_DIR}" \
  -D CMAKE_CXX_COMPILER="`which clang++`" \
  -D CMAKE_C_COMPILER="`which clang`" \
  -D CMAKE_Fortran_COMPILER="`which gfortran`" \
  -D CMAKE_CXX_FLAGS="-g -fPIC" \
  -D CMAKE_C_FLAGS="-g -fPIC" \
  -D CMAKE_Fortran_FLAGS="-g -fPIC" \
  -D CMAKE_BUILD_TYPE:STRING=DEBUG \
  -D BUILD_SHARED_LIBS:BOOL=OFF \
  -DTPL_ENABLE_MPI=OFF \
  -DTPL_ENABLE_BLAS:BOOL=ON \
   -DBLAS_LIBRARY_DIRS:FILEPATH=${OPENBLAS_ROOT}/lib \
   -DBLAS_LIBRARY_NAMES:STRING="openblas" \
  -DTPL_ENABLE_LAPACK:BOOL=ON \
   -DLAPACK_INCLUDE_DIRS:FILEPATH=${OPENBLAS_ROOT}/include \
   -DLAPACK_LIBRARY_DIRS:FILEPATH=${OPENBLAS_ROOT}/lib \
   -DLAPACK_LIBRARY_NAMES:STRING="openblas" \
  -DTrilinos_ENABLE_ALL_PACKAGES=OFF \
  -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
  -DTrilinos_ENABLE_TESTS=ON \
  -DTrilinos_MUST_FIND_ALL_TPL_LIBS=TRUE \
  -DTrilinos_ENABLE_COMPLEX=ON \
  -DTrilinos_ENABLE_OpenMP=OFF \
  -DTrilinos_ENABLE_Kokkos=ON \
  -D Kokkos_ENABLE_SERIAL=ON \
   -D Kokkos_ENABLE_TESTS=ON \
  -D Kokkos_ARCH_SKX=ON \
   -D Kokkos_ENABLE_IMPL_MDSPAN=OFF \
  -DTrilinos_ENABLE_KokkosKernels=ON \
   -D KokkosKernels_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Tpetra=ON \
   -D Tpetra_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Sacado=ON \
   -D Sacado_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Stokhos=ON \
   -D Stokhos_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Intrepid2=ON \
   -D Intrepid2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Adelus=ON \
   -D Adelus_ENABLE_TESTS:BOOL=ON \
  -DTrilinos_ENABLE_Compadre=ON \
   -D Compadre_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Panzer=ON \
   -D Panzer_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Amesos2=ON \
   -D Amesos2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Zoltan2=ON \
   -D Zoltan2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Teuchos=ON \
   -D Teuchos_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Ifpack2=ON \
   -D Ifpack2_ENABLE_TESTS=ON \
\
  -DTPL_ENABLE_Matio=OFF \
\
-DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
-DKokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
\
$TRILINOS_DIR

# Compile
make -j20
```

* The above example is the full configuration tested, but only Kokkos and Sacado likely need to be enabled above

@pzehner building with PR #7126 I am now seeing these type of errors:

```
/home/ndellin/trilinos/Trilinos/kokkos/core/src/Kokkos_CopyViews.hpp:3584:14: error: no matching function for call to 'create_mirror'
      return Kokkos::create_mirror(arg_prop, src);
             ^~~~~~~~~~~~~~~~~~~~~
/home/ndellin/trilinos/Trilinos/kokkos/core/src/Kokkos_CopyViews.hpp:3607:16: note: in instantiation of function template specialization 'Kokkos::Impl::create_mirror_view<const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>>' requested here
  return Impl::create_mirror_view(src, view_alloc());
               ^
/home/ndellin/trilinos/Trilinos/kokkos-kernels/blas/impl/KokkosBlas3_trsm_spec.hpp:108:50: note: in instantiation of function template specialization 'Kokkos::create_mirror_view<const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>>' requested here
    typename AViewType::HostMirror h_A = Kokkos::create_mirror_view(A);
                                                 ^
/home/ndellin/trilinos/Trilinos/kokkos-kernels/blas/src/KokkosBlas3_trsm.hpp:150:54: note: in instantiation of member function 'KokkosBlas::Impl::TRSM<Kokkos::Serial, Kokkos::View<const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>>, Kokkos::View<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>>, false, false>::trsm' requested here
  KokkosBlas::Impl::TRSM<execution_space, AVT, BVT>::trsm(
                                                     ^
/home/ndellin/trilinos/Trilinos/kokkos-kernels/sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp:3153:31: note: in instantiation of function template specialization 'KokkosBlas::trsm<Kokkos::Serial, Kokkos::View<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>>, Kokkos::View<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>>>' requested here
                  KokkosBlas::trsm(space, "L", "L", "N", &unit_diag, one, Ljj,
                              ^
/home/ndellin/trilinos/Trilinos/kokkos-kernels/sparse/impl/KokkosSparse_sptrsv_solve_spec.hpp:146:25: note: in instantiation of function template specialization 'KokkosSparse::Impl::Experimental::lower_tri_solve<Kokkos::Serial, KokkosSparse::Experimental::SPTRSVHandle<const unsigned long, const int, const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>>, Kokkos::Serial, Kokkos::HostSpace, Kokkos::HostSpace>, Kokkos::View<const unsigned long *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<3>>, Kokkos::View<const int *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<3>>, Kokkos::View<const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<3>>, Kokkos::View<const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<3>>, Kokkos::View<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>>>' requested here
          Experimental::lower_tri_solve(space, *sptrsv_handle, row_map, entries,
                        ^
/home/ndellin/trilinos/Trilinos/kokkos-kernels/sparse/src/KokkosSparse_sptrsv.hpp:434:26: note: in instantiation of member function 'KokkosSparse::Impl::SPTRSV_SOLVE<Kokkos::Serial, KokkosKernels::Experimental::KokkosKernelsHandle<const unsigned long, const int, const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>>, Kokkos::Serial, Kokkos::HostSpace, Kokkos::HostSpace>, Kokkos::View<const unsigned long *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<3>>, Kokkos::View<const int *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<3>>, Kokkos::View<const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<3>>, Kokkos::View<const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<3>>, Kokkos::View<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>>, false, false>::sptrsv_solve' requested here
        XType_Internal>::sptrsv_solve(space, &tmp_handle, rowmap_i, entries_i,
                         ^
/home/ndellin/trilinos/Trilinos/kokkos-kernels/sparse/src/KokkosSparse_sptrsv.hpp:472:3: note: in instantiation of function template specialization 'KokkosSparse::Experimental::sptrsv_solve<Kokkos::Serial, KokkosKernels::Experimental::KokkosKernelsHandle<const unsigned long, const int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>>, Kokkos::Serial, Kokkos::HostSpace, Kokkos::HostSpace>, Kokkos::View<const unsigned long *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0>>, Kokkos::View<int *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0>>, Kokkos::View<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutRight, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, void>, Kokkos::View<const Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0>>, Kokkos::View<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial>> *, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0>>>' requested here
  sptrsv_solve(my_exec_space, handle, rowmap, entries, values, b, x);
...
```

--- a/containers/src/Kokkos_DynRankView.hpp
+++ b/containers/src/Kokkos_DynRankView.hpp
@@ -2032,7 +2032,7 @@ inline auto create_mirror_view(
                                    T, P...>::HostMirror::data_type>::value) {
       return typename DynRankView<T, P...>::HostMirror(src);
     } else {
-      return Kokkos::create_mirror(arg_prop, src);
+      return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   } else {
     if constexpr (Impl::MirrorDRViewType<typename Impl::ViewCtorProp<
@@ -2042,7 +2042,7 @@ inline auto create_mirror_view(
           typename Impl::ViewCtorProp<ViewCtorArgs...>::memory_space, T,
           P...>::view_type(src);
     } else {
-      return Kokkos::create_mirror(arg_prop, src);
+      return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   }
 #if defined KOKKOS_COMPILER_INTEL

--- a/containers/src/Kokkos_DynamicView.hpp
+++ b/containers/src/Kokkos_DynamicView.hpp
@@ -702,7 +702,7 @@ inline auto create_mirror_view(
       return
           typename Kokkos::Experimental::DynamicView<T, P...>::HostMirror(src);
     } else {
-      return Kokkos::create_mirror(arg_prop, src);
+      return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   } else {
     if constexpr (Impl::MirrorDynamicViewType<
@@ -713,7 +713,7 @@ inline auto create_mirror_view(
           typename Impl::ViewCtorProp<ViewCtorArgs...>::memory_space, T,
           P...>::view_type(src);
     } else {
-      return Kokkos::create_mirror(arg_prop, src);
+      return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   }
 #if defined KOKKOS_COMPILER_INTEL

--- a/containers/src/Kokkos_OffsetView.hpp
+++ b/containers/src/Kokkos_OffsetView.hpp
@@ -1816,7 +1816,7 @@ namespace Impl {
 // view_alloc
 template <class T, class... P, class... ViewCtorArgs>
 inline auto create_mirror(const Kokkos::Experimental::OffsetView<T, P...>& src,
-                          const ViewCtorProp<ViewCtorArgs...>& arg_prop) {
+                          const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop) {
   check_view_ctor_args_create_mirror<ViewCtorArgs...>();
 
   if constexpr (Impl::ViewCtorProp<ViewCtorArgs...>::has_memory_space) {

--- a/containers/src/Kokkos_OffsetView.hpp
+++ b/containers/src/Kokkos_OffsetView.hpp
@@ -1816,7 +1816,7 @@ namespace Impl {
 // view_alloc
 template <class T, class... P, class... ViewCtorArgs>
 inline auto create_mirror(const Kokkos::Experimental::OffsetView<T, P...>& src,
-                          const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop) {
+                          const ViewCtorProp<ViewCtorArgs...>& arg_prop) {
   check_view_ctor_args_create_mirror<ViewCtorArgs...>();
 
   if constexpr (Impl::ViewCtorProp<ViewCtorArgs...>::has_memory_space) {
@@ -1914,7 +1914,7 @@ inline auto create_mirror_view(
       return
           typename Kokkos::Experimental::OffsetView<T, P...>::HostMirror(src);
     } else {
-      return Kokkos::create_mirror(arg_prop, src);
+      return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   } else {
     if constexpr (Impl::MirrorOffsetViewType<typename Impl::ViewCtorProp<
@@ -1924,7 +1924,7 @@ inline auto create_mirror_view(
           typename Impl::ViewCtorProp<ViewCtorArgs...>::memory_space, T,
           P...>::view_type(src);
     } else {
-      return Kokkos::create_mirror(arg_prop, src);
+      return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   }
 #if defined KOKKOS_COMPILER_INTEL

--- a/core/src/Kokkos_CopyViews.hpp
+++ b/core/src/Kokkos_CopyViews.hpp
@@ -3602,17 +3602,13 @@ inline auto create_mirror_view(
 }  // namespace Impl
 
 // public interface
-template <class T, class... P,
-          typename = std::enable_if_t<
-              std::is_void_v<typename ViewTraits<T, P...>::specialize>>>
+template <class T, class... P>
 auto create_mirror_view(const Kokkos::View<T, P...>& src) {
   return Impl::create_mirror_view(src, view_alloc());
 }
 
 // public interface that accepts a without initializing flag
-template <class T, class... P,
-          typename = std::enable_if_t<
-              std::is_void_v<typename ViewTraits<T, P...>::specialize>>>
+template <class T, class... P>
 auto create_mirror_view(Kokkos::Impl::WithoutInitializing_t wi,
                         Kokkos::View<T, P...> const& src) {
   return Impl::create_mirror_view(src, view_alloc(wi));
@@ -3621,8 +3617,7 @@ auto create_mirror_view(Kokkos::Impl::WithoutInitializing_t wi,
 // public interface that accepts a space
 template <class Space, class T, class... P,
           class Enable = std::enable_if_t<
-              Kokkos::is_space<Space>::value &&
-              std::is_void_v<typename ViewTraits<T, P...>::specialize>>>
+              Kokkos::is_space<Space>::value>>
 auto create_mirror_view(const Space&, const Kokkos::View<T, P...>& src) {
   return Impl::create_mirror_view(src,
                                   view_alloc(typename Space::memory_space()));
@@ -3631,8 +3626,7 @@ auto create_mirror_view(const Space&, const Kokkos::View<T, P...>& src) {
 // public interface that accepts a space and a without initializing flag
 template <class Space, class T, class... P,
           typename Enable = std::enable_if_t<
-              Kokkos::is_space<Space>::value &&
-              std::is_void_v<typename ViewTraits<T, P...>::specialize>>>
+              Kokkos::is_space<Space>::value>>
 auto create_mirror_view(Kokkos::Impl::WithoutInitializing_t wi, Space const&,
                         Kokkos::View<T, P...> const& src) {
   return Impl::create_mirror_view(

--- a/core/src/Kokkos_CopyViews.hpp
+++ b/core/src/Kokkos_CopyViews.hpp
@@ -3583,18 +3583,24 @@ inline auto choose_create_mirror(
       // if there are no view constructor args, call the specific public
       // function
       return Kokkos::create_mirror(src);
-    } else if constexpr (sizeof...(ViewCtorArgs) == 1 && ViewProp::has_memory_space) {
+    } else if constexpr (sizeof...(ViewCtorArgs) == 1 &&
+                         ViewProp::has_memory_space) {
       // if there is one view constructor arg and it has a memory space, call
       // the specific public function
-      return Kokkos::create_mirror(typename ViewProp::memory_space {}, src);
-    } else if constexpr (sizeof...(ViewCtorArgs) == 1 && !ViewProp::initialize) {
+      return Kokkos::create_mirror(typename ViewProp::memory_space{}, src);
+    } else if constexpr (sizeof...(ViewCtorArgs) == 1 &&
+                         !ViewProp::initialize) {
       // if there is one view constructor arg and it has a without initializing
       // mark, call the specific public function
-      return Kokkos::create_mirror(typename Kokkos::Impl::WithoutInitializing_t {}, src);
-    } else if constexpr (sizeof...(ViewCtorArgs) == 2 && ViewProp::has_memory_space && !ViewProp::initialize) {
+      return Kokkos::create_mirror(
+          typename Kokkos::Impl::WithoutInitializing_t{}, src);
+    } else if constexpr (sizeof...(ViewCtorArgs) == 2 &&
+                         ViewProp::has_memory_space && !ViewProp::initialize) {
       // if there is two view constructor args and they have a memory space and
       // a without initializing mark, call the specific public function
-      return Kokkos::create_mirror(typename Kokkos::Impl::WithoutInitializing_t {}, typename ViewProp::memory_space {}, src);
+      return Kokkos::create_mirror(
+          typename Kokkos::Impl::WithoutInitializing_t{},
+          typename ViewProp::memory_space{}, src);
     } else {
       // if there are other constructor args, call the generic public function
       return Kokkos::create_mirror(arg_prop, src);

--- a/core/src/Kokkos_CopyViews.hpp
+++ b/core/src/Kokkos_CopyViews.hpp
@@ -3564,6 +3564,48 @@ auto create_mirror(Kokkos::Impl::WithoutInitializing_t wi, Space const&,
 
 namespace Impl {
 
+template <class T, class... P, class... ViewCtorArgs>
+inline auto choose_create_mirror(
+    const Kokkos::View<T, P...>& src,
+    const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop) {
+  // Due to the fact that users can overload `Kokkos::create_mirror`, but also
+  // that they may not have implemented all of its different possible
+  // variations, this function chooses the correct private or public version of
+  // it to call.
+
+  if constexpr (std::is_void_v<typename ViewTraits<T, P...>::specialize>) {
+    // if the view is not specialized, just call the Impl function
+    return Kokkos::Impl::create_mirror(src, arg_prop);
+  } else {
+    // otherwise, recreate the public call
+    using ViewProp = Impl::ViewCtorProp<ViewCtorArgs...>;
+    if constexpr (sizeof...(ViewCtorArgs) == 0) {
+      // if there are no view constructor args, call the specific public
+      // function
+      return Kokkos::create_mirror(src);
+    } else if constexpr (sizeof...(ViewCtorArgs) == 1 && ViewProp::has_memory_space) {
+      // if there is one view constructor arg and it has a memory space, call
+      // the specific public function
+      return Kokkos::create_mirror(typename ViewProp::memory_space {}, src);
+    } else if constexpr (sizeof...(ViewCtorArgs) == 1 && !ViewProp::initialize) {
+      // if there is one view constructor arg and it has a without initializing
+      // mark, call the specific public function
+      return Kokkos::create_mirror(typename Kokkos::Impl::WithoutInitializing_t {}, src);
+    } else if constexpr (sizeof...(ViewCtorArgs) == 2 && ViewProp::has_memory_space && !ViewProp::initialize) {
+      // if there is two view constructor args and they have a memory space and
+      // a without initializing mark, call the specific public function
+      return Kokkos::create_mirror(typename Kokkos::Impl::WithoutInitializing_t {}, typename ViewProp::memory_space {}, src);
+    } else {
+      // if there are other constructor args, call the generic public function
+      return Kokkos::create_mirror(arg_prop, src);
+    }
+  }
+
+#if defined KOKKOS_COMPILER_INTEL
+  __builtin_unreachable();
+#endif
+}
+
 // create a mirror view
 // private interface that accepts arbitrary view constructor args passed by a
 // view_alloc
@@ -3581,7 +3623,7 @@ inline auto create_mirror_view(
       check_view_ctor_args_create_mirror<ViewCtorArgs...>();
       return typename Kokkos::View<T, P...>::HostMirror(src);
     } else {
-      return Kokkos::create_mirror(arg_prop, src);
+      return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   } else {
     if constexpr (Impl::MirrorViewType<typename Impl::ViewCtorProp<
@@ -3592,7 +3634,7 @@ inline auto create_mirror_view(
           typename Impl::ViewCtorProp<ViewCtorArgs...>::memory_space, T,
           P...>::view_type(src);
     } else {
-      return Kokkos::create_mirror(arg_prop, src);
+      return Kokkos::Impl::choose_create_mirror(src, arg_prop);
     }
   }
 #if defined KOKKOS_COMPILER_INTEL

--- a/core/src/Kokkos_CopyViews.hpp
+++ b/core/src/Kokkos_CopyViews.hpp
@@ -3616,8 +3616,7 @@ auto create_mirror_view(Kokkos::Impl::WithoutInitializing_t wi,
 
 // public interface that accepts a space
 template <class Space, class T, class... P,
-          class Enable = std::enable_if_t<
-              Kokkos::is_space<Space>::value>>
+          class Enable = std::enable_if_t<Kokkos::is_space<Space>::value>>
 auto create_mirror_view(const Space&, const Kokkos::View<T, P...>& src) {
   return Impl::create_mirror_view(src,
                                   view_alloc(typename Space::memory_space()));
@@ -3625,8 +3624,7 @@ auto create_mirror_view(const Space&, const Kokkos::View<T, P...>& src) {
 
 // public interface that accepts a space and a without initializing flag
 template <class Space, class T, class... P,
-          typename Enable = std::enable_if_t<
-              Kokkos::is_space<Space>::value>>
+          typename Enable = std::enable_if_t<Kokkos::is_space<Space>::value>>
 auto create_mirror_view(Kokkos::Impl::WithoutInitializing_t wi, Space const&,
                         Kokkos::View<T, P...> const& src) {
   return Impl::create_mirror_view(

--- a/core/src/Kokkos_CopyViews.hpp
+++ b/core/src/Kokkos_CopyViews.hpp
@@ -3564,46 +3564,58 @@ auto create_mirror(Kokkos::Impl::WithoutInitializing_t wi, Space const&,
 
 namespace Impl {
 
-template <class T, class... P, class... ViewCtorArgs>
+// choose a `Kokkos::create_mirror` adapted for the provided view and the
+// provided arguments
+template <class View, class... ViewCtorArgs>
 inline auto choose_create_mirror(
-    const Kokkos::View<T, P...>& src,
-    const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop) {
+    const View& src, const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop) {
   // Due to the fact that users can overload `Kokkos::create_mirror`, but also
   // that they may not have implemented all of its different possible
   // variations, this function chooses the correct private or public version of
   // it to call.
+  // This helper should be used by any overload of
+  // `Kokkos::Impl::create_mirror_view`.
 
-  if constexpr (std::is_void_v<typename ViewTraits<T, P...>::specialize>) {
+  if constexpr (std::is_void_v<typename View::traits::specialize>) {
     // if the view is not specialized, just call the Impl function
-    return Kokkos::Impl::create_mirror(src, arg_prop);
+
+    // using ADL to find the later defined overload of the function
+    using namespace Kokkos::Impl;
+
+    return create_mirror(src, arg_prop);
   } else {
     // otherwise, recreate the public call
     using ViewProp = Impl::ViewCtorProp<ViewCtorArgs...>;
+
+    // using ADL to find the later defined overload of the function
+    using namespace Kokkos;
+
     if constexpr (sizeof...(ViewCtorArgs) == 0) {
       // if there are no view constructor args, call the specific public
       // function
-      return Kokkos::create_mirror(src);
+      return create_mirror(src);
     } else if constexpr (sizeof...(ViewCtorArgs) == 1 &&
                          ViewProp::has_memory_space) {
       // if there is one view constructor arg and it has a memory space, call
       // the specific public function
-      return Kokkos::create_mirror(typename ViewProp::memory_space{}, src);
+      return create_mirror(typename ViewProp::memory_space{}, src);
     } else if constexpr (sizeof...(ViewCtorArgs) == 1 &&
                          !ViewProp::initialize) {
       // if there is one view constructor arg and it has a without initializing
       // mark, call the specific public function
-      return Kokkos::create_mirror(
-          typename Kokkos::Impl::WithoutInitializing_t{}, src);
+      return create_mirror(typename Kokkos::Impl::WithoutInitializing_t{}, src);
     } else if constexpr (sizeof...(ViewCtorArgs) == 2 &&
                          ViewProp::has_memory_space && !ViewProp::initialize) {
       // if there is two view constructor args and they have a memory space and
       // a without initializing mark, call the specific public function
-      return Kokkos::create_mirror(
-          typename Kokkos::Impl::WithoutInitializing_t{},
-          typename ViewProp::memory_space{}, src);
+      return create_mirror(typename Kokkos::Impl::WithoutInitializing_t{},
+                           typename ViewProp::memory_space{}, src);
     } else {
       // if there are other constructor args, call the generic public function
-      return Kokkos::create_mirror(arg_prop, src);
+
+      // Beware, there are some libraries using Kokkos that don't implement
+      // this overload (hence the reason for this present function to exist).
+      return create_mirror(arg_prop, src);
     }
   }
 

